staticaddr/withdraw/manager.go

```diff
@@ -649,25 +655,101 @@ func (m *Manager) createWithdrawalTx(outpoints []wire.OutPoint,
 		})
 	}
 
-	pkscript, err := txscript.PayToAddrScript(clientSweepAddress)
+	var (
+		hasChange        bool
+		dustLimit        = lnwallet.DustLimitForSize(input.P2TRSize)
+		withdrawalAmount btcutil.Amount
+		changeAmount     btcutil.Amount
+	)
+
+	// Estimate the transaction weight without change.
+	weight, err := withdrawalTxWeight(len(outpoints), withdrawAddr, false)
 	if err != nil {
-		return nil, err
+		return nil, 0, 0, err
+	}
+	feeWithoutChange := feeRate.FeeForWeight(weight)
+
+	// If the user selected a fraction of the sum of the selected deposits
+	// to withdraw, check if a change output is needed.
+	if selectedWithdrawalAmount > 0 {
+		// Estimate the transaction weight with change.
+		weight, err = withdrawalTxWeight(
+			len(outpoints), withdrawAddr, true,
+		)
+		if err != nil {
+			return nil, 0, 0, err
+		}
+		feeWithChange := feeRate.FeeForWeight(weight)
+
+		// The available change that can cover fees is the total
+		// selected deposit amount minus the selected withdrawal amount.
+		change := totalWithdrawalAmount - selectedWithdrawalAmount
+
+		switch {
+		case change-feeWithChange >= dustLimit:
+			// If the change can cover the fees without turning into
+			// dust, add a non-dust change output.
+			hasChange = true
+			changeAmount = change - feeWithChange
+			withdrawalAmount = selectedWithdrawalAmount
+
+		case change-feeWithoutChange >= 0:
+			// If the change is dust, we give it to the miners.
+			hasChange = false
+			withdrawalAmount = selectedWithdrawalAmount
+
+		default:
+			// If the fees eat into our withdrawal amount, we fail
+			// the withdrawal.
+			return nil, 0, 0, fmt.Errorf("the change doesn't " +
+				"cover for fees. Consider lowering the fee " +
+				"rate or decrease the withdrawal amount")
+		}
+	} else {
+		// If the user wants to withdraw the full amount, we don't need
+		// a change output.
+		hasChange = false
+		withdrawalAmount = totalWithdrawalAmount - feeWithoutChange
+	}
+
+	if withdrawalAmount < dustLimit {
+		return nil, 0, 0, fmt.Errorf("withdrawal amount is below " +
+			"dust limit")
+	}
+
+	if changeAmount < 0 {
+		return nil, 0, 0, fmt.Errorf("change amount is negative")
 	}
 
-	// Create the sweep output
-	sweepOutput := &wire.TxOut{
-		Value:    withdrawlOutputValue,
-		PkScript: pkscript,
+	withdrawScript, err := txscript.PayToAddrScript(withdrawAddr)
+	if err != nil {
+		return nil, 0, 0, err
 	}
 
-	msgTx.AddTxOut(sweepOutput)
+	// Create the withdrawal output.
+	msgTx.AddTxOut(&wire.TxOut{
+		Value:    int64(withdrawalAmount),
+		PkScript: withdrawScript,
+	})
+
+	if hasChange {
+		changeScript, err := txscript.PayToAddrScript(changeAddress)
```

Comment:
I think we should just get the change address here (so no need to pass it around and it is not generated if unused).

Reply:
I moved all change address data into this if-branch.